In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import matplotlib.pyplot as plt

titanic = sns.load_dataset("titanic")
titanic.drop(['embark_town','adult_male','sex','alive','class'],axis=1, inplace=True)
titanic['age'].fillna(titanic['age'].mean(),inplace=True)
new_titanic=pd.get_dummies(titanic)
new_titanic['alone']=new_titanic['alone'].replace({True:1, False:0})
new_titanic

survived  pclass        age  sibsp  parch     fare  alone  embarked_C  \
0           0       3  22.000000      1      0   7.2500      0           0   
1           1       1  38.000000      1      0  71.2833      0           1   
2           1       3  26.000000      0      0   7.9250      1           0   
3           1       1  35.000000      1      0  53.1000      0           0   
4           0       3  35.000000      0      0   8.0500      1           0   
..        ...     ...        ...    ...    ...      ...    ...         ...   
886         0       2  27.000000      0      0  13.0000      1           0   
887         1       1  19.000000      0      0  30.0000      1           0   
888         0       3  29.699118      1      2  23.4500      0           0   
889         1       1  26.000000      0      0  30.0000      1           1   
890         0       3  32.000000      0      0   7.7500      1           0   

     embarked_Q  embarked_S  who_child  who_man  who_woman  deck_A  deck_B  \
0             0           1          0        1          0       0       0   
1             0           0          0        0          1       0       0   
2             0           1          0        0          1       0       0   
3             0           1          0        0          1       0       0   
4             0           1          0        1          0       0       0   
..          ...         ...        ...      ...        ...     ...     ...   
886           0           1          0        1          0       0       0   
887           0           1          0        0          1       0       1   
888           0           1          0        0          1       0       0   
889           0           0          0        1          0       0       0   
890           1           0          0        1          0       0       0   

     deck_C  deck_D  deck_E  deck_F  deck_G  
0         0       0       0       0       0  
1         1       0       0       0       0  
2         0       0       0       0       0  
3         1       0       0       0       0  
4         0       0       0       0       0  
..      ...     ...     ...     ...     ...  
886       0       0       0       0       0  
887       0       0       0       0       0  
888       0       0       0       0       0  
889       1       0       0       0       0  
890       0       0       0       0       0  

[891 rows x 20 columns]

In [ ]:
from sklearn.preprocessing import StandardScaler

norm_scaler = StandardScaler()
new_titanic['age'] = norm_scaler.fit_transform(new_titanic[['age']])
new_titanic['fare'] = norm_scaler.fit_transform(new_titanic[['fare']])
new_titanic

survived  pclass       age  sibsp  parch      fare  alone  embarked_C  \
0           0       3 -0.592481      1      0 -0.502445      0           0   
1           1       1  0.638789      1      0  0.786845      0           1   
2           1       3 -0.284663      0      0 -0.488854      1           0   
3           1       1  0.407926      1      0  0.420730      0           0   
4           0       3  0.407926      0      0 -0.486337      1           0   
..        ...     ...       ...    ...    ...       ...    ...         ...   
886         0       2 -0.207709      0      0 -0.386671      1           0   
887         1       1 -0.823344      0      0 -0.044381      1           0   
888         0       3  0.000000      1      2 -0.176263      0           0   
889         1       1 -0.284663      0      0 -0.044381      1           1   
890         0       3  0.177063      0      0 -0.492378      1           0   

     embarked_Q  embarked_S  who_child  who_man  who_woman  deck_A  deck_B  \
0             0           1          0        1          0       0       0   
1             0           0          0        0          1       0       0   
2             0           1          0        0          1       0       0   
3             0           1          0        0          1       0       0   
4             0           1          0        1          0       0       0   
..          ...         ...        ...      ...        ...     ...     ...   
886           0           1          0        1          0       0       0   
887           0           1          0        0          1       0       1   
888           0           1          0        0          1       0       0   
889           0           0          0        1          0       0       0   
890           1           0          0        1          0       0       0   

     deck_C  deck_D  deck_E  deck_F  deck_G  
0         0       0       0       0       0  
1         1       0       0       0       0  
2         0       0       0       0       0  
3         1       0       0       0       0  
4         0       0       0       0       0  
..      ...     ...     ...     ...     ...  
886       0       0       0       0       0  
887       0       0       0       0       0  
888       0       0       0       0       0  
889       1       0       0       0       0  
890       0       0       0       0       0  

[891 rows x 20 columns]

In [ ]:
import sklearn
from sklearn.model_selection import train_test_split
X = new_titanic.drop(['survived'], axis=1)
y = new_titanic['survived']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=42)
y_train

331    0
733    0
382    0
704    0
813    0
      ..
106    1
270    0
860    0
435    1
102    0
Name: survived, Length: 712, dtype: int64

In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(solver='liblinear', max_iter=1000)
model.fit(X_train, y_train)

y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix

cm_train = confusion_matrix(y_train, y_train_pred)
cm_train

array([[393,  51],
       [ 64, 204]])

In [ ]:
accuracy=(393+204)/(393+51+204+64)
precision=204/(204+51)
recall=204/(204+64)
f1=2*(precision*recall)/(precision+recall)
print("Train accuracy, precision,recall and f1:", accuracy*100, precision*100, recall*100,f1*100)

Train accuracy, precision,recall and f1: 83.84831460674157 80.0 76.11940298507463 78.01147227533461


In [ ]:
cm_test = confusion_matrix(y_test, y_test_pred)
cm_test

array([[90, 15],
       [18, 56]])

In [ ]:
accuracy_test=(90+56)/(90+56+15+18)
precision_test=56/(56+15)
recall_test=56/(56+18)
f1_test=2*(precision_test*recall_test)/(precision_test+recall_test)
print("Test accuracy, precision,recall and f1:", accuracy_test*100, precision_test*100, recall_test*100,f1_test*100)

Test accuracy, precision,recall and f1: 81.56424581005587 78.87323943661971 75.67567567567568 77.24137931034481


In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

train_accuracy = accuracy_score(y_train, y_train_pred)
test_accuracy = accuracy_score(y_test, y_test_pred)
print("Train accuracy of the model is %.2f percent"%(train_accuracy*100))
print("Test accuracy of the model is %.2f percent"%(test_accuracy*100))

train_precision = precision_score(y_train, y_train_pred)
test_precision = precision_score(y_test, y_test_pred)
print("Train precision: %.2f"%(train_precision*100))
print("Test precision: %.2f"%(test_precision*100))

train_recall = recall_score(y_train, y_train_pred)
test_recall = recall_score(y_test, y_test_pred)
print("Train recall: %.2f"%(train_recall*100))
print("Test recall: %.2f"%(test_recall*100))

train_f1 = f1_score(y_train, y_train_pred)
test_f1 = f1_score(y_test, y_test_pred)
print("Train f1: %.2f"%(train_f1*100))
print("Test f1: %.2f"%(test_f1*100))

train_roc_auc = roc_auc_score(y_train, y_train_pred)
test_roc_auc = roc_auc_score(y_test, y_test_pred)
print("Train roc_auc: %.2f"%(train_roc_auc*100))
print("Test roc_auc: %.2f"%(test_roc_auc*100))

Train accuracy of the model is 83.85 percent
Test accuracy of the model is 81.56 percent
Train precision: 80.00
Test precision: 78.87
Train recall: 76.12
Test recall: 75.68
Train f1: 78.01
Test f1: 77.24
Train roc_auc: 82.32
Test roc_auc: 80.69



Train accuracy, precision,recall and f1: 84.08450704225352 80.0 76.11940298507463 78.01147227533461

Test accuracy, precision,recall and f1: 81.56424581005587 78.87323943661971 75.67567567567568 77.24137931034481

In [ ]:
from sklearn.tree import DecisionTreeClassifier

tree_model = DecisionTreeClassifier().fit(X_train, y_train)

y_train_pred = tree_model.predict(X_train)
y_test_pred = tree_model.predict(X_test)

print(confusion_matrix(y_train, y_train_pred))
print(confusion_matrix(y_test, y_test_pred))

print("------")
print("Train accuracy: %.2f"%(accuracy_score(y_train, y_train_pred)*100))
print("Test accuracy: %.2f"%(accuracy_score(y_test, y_test_pred)*100))
print("------")
print("Train F1: %.2f"%(f1_score(y_train, y_train_pred)*100))
print("Test F1: %.2f"%(f1_score(y_test, y_test_pred)*100))

[[443   1]
 [ 10 258]]
[[76 29]
 [19 55]]
------
Train accuracy: 98.46
Test accuracy: 73.18
------
Train F1: 97.91
Test F1: 69.62


In [ ]:
from sklearn.ensemble import RandomForestClassifier

forest_model = RandomForestClassifier().fit(X_train, y_train)

y_train_pred = forest_model.predict(X_train)
y_test_pred = forest_model.predict(X_test)

print(confusion_matrix(y_train, y_train_pred))
print(confusion_matrix(y_test, y_test_pred))

print("------")
print("Train accuracy: %.2f"%(accuracy_score(y_train, y_train_pred)*100))
print("Test accuracy: %.2f"%(accuracy_score(y_test, y_test_pred)*100))
print("------")
print("Train F1: %.2f"%(f1_score(y_train, y_train_pred)*100))
print("Test F1: %.2f"%(f1_score(y_test, y_test_pred)*100))

[[442   2]
 [  9 259]]
[[91 14]
 [21 53]]
------
Train accuracy: 98.46
Test accuracy: 80.45
------
Train F1: 97.92
Test F1: 75.18


In [ ]:
from sklearn import svm
from sklearn.datasets import make_classification
clf = svm.SVC(kernel='linear', C=1.0)
clf.fit(X_train, y_train)
y_train_pred=clf.predict(X_train)
y_test_pred=clf.predict(X_test)
y_test_pred


print(confusion_matrix(y_train, y_train_pred))
print(confusion_matrix(y_test, y_test_pred))

print("------")
print("Train accuracy: %.2f"%(accuracy_score(y_train, y_train_pred)*100))
print("Test accuracy: %.2f"%(accuracy_score(y_test, y_test_pred)*100))
print("------")
print("Train F1: %.2f"%(f1_score(y_train, y_train_pred)*100))
print("Test F1: %.2f"%(f1_score(y_test, y_test_pred)*100))

[[391  53]
 [ 68 200]]
[[89 16]
 [19 55]]
------
Train accuracy: 83.01
Test accuracy: 80.45
------
Train F1: 76.78
Test F1: 75.86


**Tree model**
Train accuracy: 98.46,
Test accuracy: 73.74,
Train F1: 97.91,
Test F1: 69.68.

**Forest model**
Train accuracy: 98.46,
Test accuracy: 79.33,
Train F1: 97.91,
Test F1: 73.76

**LogREg model**
Train accuracy:83.57,
Test accuracy: 81.01,
Train f1: 77.71,
Test f1: 76.71

**SVC model**
Train accuracy: 83.01,
Test accuracy: 80.45,
Train F1: 76.78,
Test F1: 75.86

As we can see, random forest and tree models are overfitted since the train metrics are very high, while test models are much less. Logistic regression and SVC models are balanced. Logistic regression has slightly higher results, therefore I will chose it.

k-fold evaluation below also shows that robustly logistic regression is the best model

In [ ]:
from sklearn.model_selection import cross_val_score
cv_scores = cross_val_score(model, X, y, cv=5)
print("Cross-Validation Scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

from sklearn.model_selection import cross_val_score
cv_scores = cross_val_score(tree_model, X, y, cv=5)
print("Cross-Validation Scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

from sklearn.model_selection import cross_val_score
cv_scores = cross_val_score(forest_model, X, y, cv=5)
print("Cross-Validation Scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

from sklearn.model_selection import cross_val_score
cv_scores = cross_val_score(clf, X, y, cv=5)
print("Cross-Validation Scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

Cross-Validation Scores: [0.81005587 0.82022472 0.80337079 0.82022472 0.85955056]
Mean CV Score: 0.8226853304877284
Cross-Validation Scores: [0.75418994 0.76404494 0.80898876 0.80898876 0.78651685]
Mean CV Score: 0.784545853995355
Cross-Validation Scores: [0.7877095  0.79775281 0.81460674 0.75842697 0.84269663]
Mean CV Score: 0.8002385286548238
Cross-Validation Scores: [0.81564246 0.80898876 0.81460674 0.79213483 0.84269663]
Mean CV Score: 0.8148138848785387


In [ ]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import uniform

param_dist = {
    'penalty': ['l1','l2'],
    'C': uniform(0.001, 100),
    'max_iter': [50, 100, 200],
    'class_weight': [None, 'balanced']
}
model = LogisticRegression(solver='liblinear', max_iter=1000)
model.fit(X_train, y_train)

random_search = RandomizedSearchCV(model, param_dist, n_iter=10, cv=5, error_score='raise')
random_search.fit(X_train, y_train)

best_params_random = random_search.best_params_
best_model_random = random_search.best_estimator_

test_accuracy_random = best_model_random.score(X_test, y_test)
y_pred = best_model_random.predict(X_test)
f1 = f1_score(y_test, y_pred)
roc_auc=roc_auc_score(y_test,y_pred)

print("Best Hyperparameters (Random Search):", best_params_random)
print("Test Accuracy (Random Search):", test_accuracy_random)
print("Test f1 (Random Search):", f1)
print("Test roc_auc (Random Search):",roc_auc)

Best Hyperparameters (Random Search): {'C': 5.106143582209304, 'class_weight': None, 'max_iter': 200, 'penalty': 'l2'}
Test Accuracy (Random Search): 0.8100558659217877
Test f1 (Random Search): 0.7671232876712328
Test roc_auc (Random Search): 0.8021879021879021


Original Test accuracy, precision,recall and f1: 81.56424581005587 78.87323943661971 75.67567567567568 77.24137931034481
Train roc_auc: 82.32
Test roc_auc: 80.69

Тренируя модель на новых гиперпараметрах, я получила примерно такие же метрики как в оригинальной модели. Это связано с тем что параметры были выбраны рандомно.

In [ ]:
from sklearn.model_selection import GridSearchCV
model = LogisticRegression(solver='liblinear', max_iter=1000)
model.fit(X_train, y_train)

param_grid = {
    'penalty': ['l1', 'l2'],
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'max_iter': [50, 100, 200],
    'class_weight': [None, 'balanced']
}

grid_search = GridSearchCV(model, param_grid, cv=5)
grid_search.fit(X_train, y_train)

best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

y_pred = best_model_random.predict(X_test)

test_accuracy = accuracy_score(y_test,y_pred)
f1 = f1_score(y_test, y_pred)
roc_auc=roc_auc_score(y_test,y_pred)
print("Best Hyperparameters (Grid Search):", best_params)
print("Test Accuracy (Grid Search):", test_accuracy)
print("Test f1 (Random Search):", f1)
print("Test roc_auc (Random Search):",roc_auc)

/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Best Hyperparameters (Grid Search): {'C': 1, 'class_weight': None, 'max_iter': 50, 'penalty': 'l1'}
Test Accuracy (Grid Search): 0.8100558659217877
Test f1 (Random Search): 0.7671232876712328
Test roc_auc (Random Search): 0.8021879021879021


Original Test accuracy, precision,recall and f1: 81.56424581005587 78.87323943661971 75.67567567567568 77.24137931034481 Train roc_auc: 82.32 Test roc_auc: 80.69